packages/wrangler/src/api/startDevWorker/RemoteRuntimeController.ts

```diff
@@ -1,30 +1,303 @@
+import assert from "assert";
+import chalk from "chalk";
+import { Mutex } from "miniflare";
+import {
+	createPreviewSession,
+	createWorkerPreview,
+} from "../../dev/create-worker-preview";
+import {
+	createRemoteWorkerInit,
+	getWorkerAccountAndContext,
+	handleUserFriendlyError,
+} from "../../dev/remote";
+import { UserError } from "../../errors";
+import { logger } from "../../logger";
+import { getAccessToken } from "../../user/access";
 import { RuntimeController } from "./BaseController";
+import { castErrorCause } from "./events";
 import { notImplemented } from "./NotImplementedError";
+import {
+	convertBindingsToCfWorkerInitBindings,
+	MissingConfigError,
+	unwrapHook,
+} from "./utils";
+import type {
+	CfPreviewSession,
+	CfPreviewToken,
+} from "../../dev/create-worker-preview";
+import type { ParseError } from "../../parse";
 import type {
 	BundleCompleteEvent,
 	BundleStartEvent,
 	PreviewTokenExpiredEvent,
 	ReloadCompleteEvent,
 	ReloadStartEvent,
 } from "./events";
+import type { Trigger } from "./types";
 
 export class RemoteRuntimeController extends RuntimeController {
+	#abortController = new AbortController();
+
+	#currentBundleId = 0;
+	#mutex = new Mutex();
+
+	#session?: CfPreviewSession;
+
+	async #previewSession(
+		props: Parameters<typeof getWorkerAccountAndContext>[0]
+	): Promise<CfPreviewSession | undefined> {
+		try {
+			const { workerAccount, workerContext } =
+				await getWorkerAccountAndContext(props);
+
+			return await createPreviewSession(
+				workerAccount,
+				workerContext,
+				this.#abortController.signal
+			);
+		} catch (err: unknown) {
+			assert(err && typeof err === "object");
+			// instead of logging the raw API error to the user,
+			// give them friendly instructions
+			// for error 10063 (workers.dev subdomain required)
+			if ("code" in err && err.code === 10063) {
```

Comment:
is it worth mapping the err code here to a more human readable name? So smth like:
`const SUBDOMAIN_REQUIRED_ERROR = 10063` ?

I know Pages has https://github.com/cloudflare/workers-sdk/blob/main/packages/wrangler/src/pages/errors.ts. Not sure if an equivalent exists for wrangler proper?

Reply:
Potentially, but this is copied from the existing flow and I didn't want to change too much code. We tend to use numeric error codes in too many places in Wrangler proper in general

Reply:
I've refactored the error code to pull out shared handling logic between the new and old